src/QtAVPlayer/qavvideoframe.cpp

```diff
@@ -40,8 +40,8 @@ class QAVVideoFramePrivate : public QAVFramePrivate
     QAVVideoFramePrivate(QAVVideoFrame *q) : q_ptr(q) { }
 
     QAVVideoBuffer &videoBuffer() const
-    {
-        if (!buffer) {
+    {       
+        if (!buffer || q_ptr->frame() != buffer->frame().frame()) {
             auto c = videoCodec(stream.codec().data());
             auto buf = c && c->device() && frame->format == c->device()->format() ? c->device()->videoBuffer(*q_ptr) : new QAVVideoBuffer_CPU(*q_ptr);
```

Comment:
or it is both video frames with the same format. But the content is not changed, since the buffer keeps old data? How to reproduce it? Can you find a simple test case?

Reply:
```
QObject::connect(&player, &QAVPlayer::videoFrame, &a, [&](const QAVVideoFrame& frame) {
		QVideoFrame videoFrame = frame.convertTo(AVPixelFormat::AV_PIX_FMT_YUV420P) ;	
	},Qt::DirectConnection);
```

Reply:
When use Qt::QueuedConnection,  `emit QAVPlayer::videoFrame` construct a QAVVideoFrame and send it to queue,
call QAVVideoFrame::convertTo function have no effect on `QAVVideoFrame frame;` in QAVPlayerPrivate::doPlayVideo()

When use Qt::DirectConnection,the  `QAVVideoFrame frame;` in  QAVPlayerPrivate::doPlayVideo() call convertTo  and create QAVVideoBuffer_GPU,at the same time QAVVideoBuffer_GPU also store AVFrame pointer

Reply:
The better solution may be
```
    QAVVideoBuffer &videoBuffer() const
    {
        if (!buffer) {
            auto c = videoCodec(stream.codec().data());
            auto buf = c && c->device() && frame->format == c->device()->format() ? c->device()->videoBuffer(*q_ptr) : new QAVVideoBuffer_CPU(*q_ptr);
            const_cast<QAVVideoFramePrivate*>(this)->buffer.reset(buf);
        }
        else 
            const_cast<QAVVideoFrame&>(buffer->frame()) = *q_ptr;
        
        return *buffer;
    }
```

Reply:
thanks, let's begin from the unit test which reproduces it

Reply:
ok, reproduced but sorry, the bug seems was due to it tried to transfer data twice, not because it keeps a ref. Also do you know that to render the frames, you _must_ submit the frame in guy thread? Means direct connection is useless unless you want to handle this yourself?
https://github.com/valbok/QtAVPlayer/pull/266